### ISSUE TITLE
ODataUnrecognizedPathException is thrown for non-odata routes when odata route prefix is null or empty

### ISSUE THREAD

**Issue by ud09** [OData/WebApi#2324]
### Assemblies affected

Microsoft.AspNetCore.OData 7.5.0

### Reproduce steps
Startup.cs

    using System.Linq;
    using Microsoft.AspNet.OData.Builder;
    using Microsoft.AspNet.OData.Extensions;
    using Microsoft.AspNetCore.Builder;
    using Microsoft.AspNetCore.Hosting;
    using Microsoft.Extensions.Configuration;
    using Microsoft.Extensions.DependencyInjection;
    using Microsoft.Extensions.Hosting;
    using Microsoft.OData.Edm;
    
    namespace WeatherAPI
    {
        public class Startup
        {
            public Startup(IConfiguration configuration)
            {
                Configuration = configuration;
            }
    
            public IConfiguration Configuration { get; }
    
            public void ConfigureServices(IServiceCollection services)
            {
                services.AddControllers();
                services.AddOData();
            }
    
            public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
            {
                if (env.IsDevelopment())
                {
                    app.UseDeveloperExceptionPage();
                }
    
                app.UseHttpsRedirection();
    
                app.UseRouting();
    
                app.UseAuthorization();
    
                app.UseEndpoints(endpoints =>
                {
                    endpoints.MapControllers();
                    endpoints.Select().Filter().OrderBy().Count().MaxTop(10);
                    endpoints.MapODataRoute("odata", null, GetEdmModel());
                });
            }
    
            private IEdmModel GetEdmModel()
            {
                var odataBuilder = new ODataConventionModelBuilder();
                odataBuilder.EntitySet<WeatherForecast>("WeatherForecast");
                return odataBuilder.GetEdmModel();
            }
        }
    }
    
WeatherForecastController.cs

    using System;
    using System.Collections.Generic;
    using System.Linq;
    using Microsoft.AspNet.OData;
    using Microsoft.AspNetCore.Mvc;
    namespace WeatherAPI.Controllers
    {
        public class WeatherForecastController : ControllerBase
        {
            private static readonly string[] Summaries = new[]
            {
                "Freezing", "Bracing", "Chilly", "Cool", "Mild", "Warm", "Balmy", "Hot", "Sweltering", "Scorching"
            };
    
            [HttpGet]
            [Route("Test")]
            public IActionResult Test()
            {
                return Ok("My Test");
            }
    
            [EnableQuery]
            public IEnumerable<WeatherForecast> Get()
            {
                var rng = new Random();
                return Enumerable.Range(1, 5).Select(index => new WeatherForecast
                {
                    Id = Guid.NewGuid(),
                    Date = DateTime.Now.AddDays(index),
                    TemperatureC = rng.Next(-20, 55),
                    Summary = Summaries[rng.Next(Summaries.Length)]
                })
                .ToArray();
            }
        }
    }

### Expected result

Here I give null to the odata route prefix since it was also null in a webapi that I'm trying to convert to aspnet.core. Better not to change the prefix.

    endpoints.MapODataRoute("odata", null, GetEdmModel());

and 
http://localhost:5001/test  request throws ODataUnrecognizedPathException (Resource not found for the segment 'test'). 
Later it's caught somewhere in the odata call stack , but **I'd like to learn if there is a way to bypass odata endpoint route matching so this exception never occurs for the non-odata routes. Or is this the expected behavior ?** 

### Additional detail

>	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataPathParser.CreateDynamicPathSegment(Microsoft.OData.UriParser.ODataPathSegment previous, string identifier, string parenthesisExpression)	Unknown
 	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataPathParser.CreateFirstSegment(string segmentText)	Unknown
 	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataPathParser.ParsePath(System.Collections.Generic.ICollection<string> segments)	Unknown
 	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataPathFactory.BindPath(System.Collections.Generic.ICollection<string> segments, Microsoft.OData.UriParser.ODataUriParserConfiguration configuration)	Unknown
 	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataUriParser.ParsePathImplementation()	Unknown
 	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataUriParser.Initialize()	Unknown
 	Microsoft.OData.Core.dll!Microsoft.OData.UriParser.ODataUriParser.ParsePath()	Unknown
 	Microsoft.AspNetCore.OData.dll!Microsoft.AspNet.OData.Routing.DefaultODataPathHandler.Parse(string serviceRoot, string odataPath, System.IServiceProvider requestContainer, bool template)	Unknown
 	Microsoft.AspNetCore.OData.dll!Microsoft.AspNet.OData.Routing.DefaultODataPathHandler.Parse(string serviceRoot, string odataPath, System.IServiceProvider requestContainer)	Unknown
 	Microsoft.AspNetCore.OData.dll!Microsoft.AspNet.OData.Routing.ODataPathRouteConstraint.GetODataPath(string oDataPathString, string uriPathString, string queryString, System.Func<System.IServiceProvider> requestContainerFactory)	Unknown
 	Microsoft.AspNetCore.OData.dll!Microsoft.AspNet.OData.Extensions.ODataEndpointRouteValueTransformer.TransformAsync(Microsoft.AspNetCore.Http.HttpContext httpContext, Microsoft.AspNetCore.Routing.RouteValueDictionary values)	Unknown
 	Microsoft.AspNetCore.Mvc.Core.dll!Microsoft.AspNetCore.Mvc.Routing.DynamicControllerEndpointMatcherPolicy.ApplyAsync(Microsoft.AspNetCore.Http.HttpContext httpContext, Microsoft.AspNetCore.Routing.Matching.CandidateSet candidates)	Unknown
 	Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.Matching.DfaMatcher.SelectEndpointWithPoliciesAsync(Microsoft.AspNetCore.Http.HttpContext httpContext, Microsoft.AspNetCore.Routing.Matching.IEndpointSelectorPolicy[] policies, Microsoft.AspNetCore.Routing.Matching.CandidateSet candidateSet)	Unknown
 	Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.Matching.DfaMatcher.MatchAsync(Microsoft.AspNetCore.Http.HttpContext httpContext)	Unknown
 	Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.Matching.DataSourceDependentMatcher.MatchAsync(Microsoft.AspNetCore.Http.HttpContext httpContext)	Unknown
 	Microsoft.AspNetCore.Routing.dll!Microsoft.AspNetCore.Routing.EndpointRoutingMiddleware.Invoke(Microsoft.AspNetCore.Http.HttpContext httpContext)	Unknown

**Comment by gathogojr** [comment]
Hi @ud09 Thank you for reaching out to us. I tried to reproduce the issue that you described using `Microsoft.AspNetCore.OData 7.5.0` plus the exact code you provided in the **Reproduce steps** section - the only thing I inferred is the `WeatherForecast` class. I could successfully send requests to http://localhost:5001/test (_different port for my case_). Do you think you can you share a link to a repro? You could also mention the version of ASP.NET Core you're using.

**Comment by ud09** [comment]
Hi @gathogojr , did you set the odata route prefix to null ? Did you turn on all clr exceptions in your ide ? As i said the first chance exception is thrown in microsoft.odata.core as shown in the callstack above , but also caught somewhere in the odata code flow.(but i don’t know where) Is there a way to tell odata not to try to match the route for non odata routes when prefix is null , so that this code flow is not run ? Hope you can reproduce, thanks.

**Comment by gathogojr** [comment]
> did you set the odata route prefix to null ?

Of course. Like I mentioned, I copied the exact code you provided in the Reproduce steps section - the only thing I inferred is the `WeatherForecast` class.

### Service Configuration
```c#
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseHttpsRedirection();

            app.UseRouting();

            app.UseAuthorization();

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapControllers();
                endpoints.Select().Filter().OrderBy().Count().MaxTop(10);
                endpoints.MapODataRoute("odata", null, GetEdmModel());
            });
        }

        private IEdmModel GetEdmModel()
        {
            var odataBuilder = new ODataConventionModelBuilder();
            odataBuilder.EntitySet<WeatherForecast>("WeatherForecast");
            return odataBuilder.GetEdmModel();
        }
```

![image](https://user-images.githubusercontent.com/28291332/97487307-c27c7c80-196d-11eb-94b6-8ca30a6b9588.png)

I can't reproduce with the information you provided so you'll have to give me a little more to go with.

**Comment by ud09** [comment]
Did you turn on all clr exceptions and run in debug mode ? It should break on this exception before the request reaches action method. It’s a first chance exception and doesn’t cause any failure.

**Comment by gathogojr** [comment]
![image](https://user-images.githubusercontent.com/28291332/97529908-04371280-19c2-11eb-9bec-a736a9f5282d.png)

Still, no `ODataUnrecognizedPathException` observed...

**Comment by ud09** [comment]
[https://github.com/ud09/ODataWithEndpointRouting3.1](https://github.com/ud09/ODataWithEndpointRouting3.1)
Hello,  this is the fork I've made from Hassan Habbib's sample. I've updated to Microsoft.AspNetCore.OData 7.5.1 , added Test method and set odata prefix to null. When I start debugging with all exceptions on, the exception is thrown. Thanks again.
![image](https://user-images.githubusercontent.com/6473521/97662712-d9fc5800-1a88-11eb-963e-1ff504e5c1c8.png)
